src/main/java/dev/openfeature/sdk/providers/memory/InMemoryProvider.java

```diff
@@ -87,68 +90,52 @@ public void updateFlag(String flagKey, Flag<?> flag) {
     @Override
     public ProviderEvaluation<Boolean> getBooleanEvaluation(String key, Boolean defaultValue,
             EvaluationContext evaluationContext) {
-        return getEvaluation(key, defaultValue, evaluationContext, Boolean.class);
+        return getEvaluation(key, evaluationContext, Boolean.class);
     }
 
     @Override
     public ProviderEvaluation<String> getStringEvaluation(String key, String defaultValue,
             EvaluationContext evaluationContext) {
-        return getEvaluation(key, defaultValue, evaluationContext, String.class);
+        return getEvaluation(key, evaluationContext, String.class);
     }
 
     @Override
     public ProviderEvaluation<Integer> getIntegerEvaluation(String key, Integer defaultValue,
             EvaluationContext evaluationContext) {
-        return getEvaluation(key, defaultValue, evaluationContext, Integer.class);
+        return getEvaluation(key, evaluationContext, Integer.class);
     }
 
     @Override
     public ProviderEvaluation<Double> getDoubleEvaluation(String key, Double defaultValue,
             EvaluationContext evaluationContext) {
-        return getEvaluation(key, defaultValue, evaluationContext, Double.class);
+        return getEvaluation(key, evaluationContext, Double.class);
     }
 
     @SneakyThrows
     @Override
     public ProviderEvaluation<Value> getObjectEvaluation(String key, Value defaultValue,
             EvaluationContext evaluationContext) {
-        return getEvaluation(key, defaultValue, evaluationContext, Value.class);
+        return getEvaluation(key, evaluationContext, Value.class);
     }
 
     private <T> ProviderEvaluation<T> getEvaluation(
-            String key, T defaultValue, EvaluationContext evaluationContext, Class<?> expectedType
+            String key, EvaluationContext evaluationContext, Class<?> expectedType
     ) throws OpenFeatureError {
         if (!ProviderState.READY.equals(state)) {
-            ErrorCode errorCode = ErrorCode.PROVIDER_NOT_READY;
-            if (ProviderState.ERROR.equals(state)) {
-                errorCode = ErrorCode.GENERAL;
+            if (ProviderState.NOT_READY.equals(state)) {
+                throw new ProviderNotReadyError("provider not yet initialized");
```

Comment:
@liran2000 throwing like this ends in the same result, but it's a bit easier to trace, and also simpler to write. This is how we generally recommend it's done. There is a small performance cost in serialization the exception, but I think it's worth it.